### PR TITLE
feat: Added `ssgParams` middleware

### PR DESCRIPTION
**Pull request by nakasyou**
I added `ssgParams` middleware.
This is implemented https://github.com/honojs/hono/pull/1904#issuecomment-188809839.

There is 2 changes.

#### 1. API like `generateStaticPaths`.
You can use API like `generateStaticPaths`.
Example:
```tsx
import { Hono } from 'hono'
import { toSSG, ssgParams } from 'hono/ssg'
import * as fs from 'fs/promises'

app.get('/', c => c.html(<h1>Top page</h1>))
app.get('/post/:id', ssgParams([{ id: '1' }]), c => c.html(<h1>{c.req.params('id')}</h1>)

await toSSG(app, fs)
```
#### 2. Switching SSR/SSG
You can add No-SSG routes such as:
```tsx
import { Hono } from 'hono'
import { toSSG, ssgParams } from 'hono/ssg'
import * as fs from 'fs/promises'

app.get('/', c => c.html(<h1>Top page</h1>)) // SSG
app.get(
  '/api',
  ssgParams(false), // Don't SSG
  c => c.json({ status: 'ok' })
) 

await toSSG(app, fs)
```
This code is SSG-based mode like [Astro hybrid mode](https://docs.astro.build/en/guides/server-side-rendering/#enable-on-demand-server-rendering).

Not only this code, you can use SSR-Based mode.
```tsx
import { Hono } from 'hono'
import { toSSG, ssgParams } from 'hono/ssg'
import * as fs from 'fs/promises'

app.get('/',
  ssgParams(true), // SSG
  c => c.html(<h1>Top page</h1>)
) // SSG
// Or, this code also works.
app.get('/post/:id', ssgParams([{ id: '1' }]), c => c.html(<h1>{c.req.params('id')}</h1>))

app.get(
  '/api',
  c => c.json({ status: 'ok' })
) // This route will not generate SSG files. `toSSG` don't generate route if not explicitly specified.

await toSSG(app, fs, { default: 'ssr' })
```

Hereby, you can write code to combine both SSG and SSR such as:
```tsx
// dev.tsx
import { Hono } from 'hono'
import { ssgParams } from 'hono/ssg'

const app = new Hono()

app.get('/', ssgParams(true), c => c.html(<h1>Hello world!</h1>)) // Top page is static.
app.get('/api/random', c => c.json({ result: Math.random() })) // API is dynamic.

export default app
```
```tsx
// prediction.ts
import { Hono } from 'hono'
import app from './dev'

const mainApp = new Hono()
mainApp.use('*', serveStatic({ root: 'static' }))
mainApp.route('/', app)

export default app
```
```tsx
// build.ts
import { Hono } from 'hono'
import app from './dev'
import { toSSG, ssgParams } from 'hono/ssg'
import * as fs from 'fs/promises'

toSSG(app, fs, { default: 'ssr' })
```
### Author should do the followings, if applicable

- [x] Add tests
- [x] Run tests
- [x] `yarn denoify` to generate files for Deno

### CODE REVIEW

**Comment by yusukebe** [comment]
@nakasyou 

Seems to be good. If it's ready for reviewing, please let me know!

**Comment by watany-dev** [comment]
I think it would be better to separate the generateStaticPaths and the hybrid SSR/SSG middleware. At least to me, the interface seems too complex.

For the latter middleware, it might be good to simplify by just using a flag like isSSG() or isSSG/isSSR. However, I haven't read the underlying implementation of Astro, so these opinions might not be correct:)

**Comment by yusukebe** [comment]
Hi @nakasyou 

For this PR, I created a simple application and used this feature before reviewing the code.

Repo: https://github.com/yusukebe/hono-ssg-example/tree/ramen
Demo: https://3282047d.hono-ssg.pages.dev/

Feeling of using `ssgParams` was great! For example, here is the code:

```tsx
app.get(
  '/shops/:id',
  ssgParams(async () => {
    const shops = await getShops()
    return shops.map((shop) => ({ id: shop.id }))
  }),
  async (c) => {
    const shop = await getShop(c.req.param('id'))
    if (!shop) {
      return c.notFound()
    }
    return c.render(
      <div>
        <h1>{shop.name}</h1>
      </div>
    )
  }
)
```

However, as @watany-dev says, it seems complicated to include "hybrid SSR/SSG middleware" things in this feature. It seems unnatural to accept a boolean as an argument to `ssgParams()`.

I think it would be good to implement only the "generateStaticPaths function" in this PR, and propose SSG/SSR switching in another PR.

File: src/helper/ssg/index.ts [hunk 1]
```diff
@@ -19,24 +20,90 @@ const generateFilePath = (routePath: string, outDir: string) => {
   return path.join(outDir, fileName)
 }
 
+interface SSGParam {
+  [key: string]: string
+}
+type SSGParams = SSGParam[]
+interface SSGParamsMiddleware {
+  (generateParams: () => SSGParams | Promise<SSGParams>): MiddlewareHandler
```
Comment (yusukebe) [inline src/helper/ssg/index.ts:28]: How about passing an instance of Context?

```ts
(generateParams: (c: Context) => SSGParams | Promise<SSGParams>): MiddlewareHandler
```

By doing this, you can write cool code like the following:

```ts
app.get(
  '/shops',
  createMiddleware<{ Variables: { shops: Shop[] } }>(async (c) => {
    const shops = await getShops()
    c.set('shops', shops)
  }),
  ssgParams(async (c) => {
    return c.var.shops.map((shop) => ({ id: shop.id }))
  }),
  async (c) => {
    return c.render(
      <div>
        <h1>ラーメン屋さん</h1>
        <ul>
          {c.var.shops.map((shop) => {
            return (
              <li>
                <a href={`/shops/${shop.id}`}>{shop.name}</a>
              </li>
            )
          })}
        </ul>
      </div>
    )
  }
)
```

Comment (nakasyou) [reply src/helper/ssg/index.ts:28]: It's really cool idea!
I'll implement it.

**Comment by nakasyou** [comment]
@watany-dev @yusukebe 
> I think it would be better to separate the generateStaticPaths and the hybrid SSR/SSG middleware. At least to me, the interface seems too complex.

> However, as @watany-dev says, it seems complicated to include "hybrid SSR/SSG middleware" things in this feature. It seems unnatural to accept a boolean as an argument to ssgParams().

I think so after I thought again.
Certainly, to separate staticPaths API and hybrid API is quite understandable.
So I'll make other PR.

File: src/helper/ssg/index.ts (hunk 1 above)
Comment (yusukebe) [reply src/helper/ssg/index.ts]: Thank you. We might need to change the type, but I'll do it after this PR is merged!

**Comment by nakasyou** [comment]
@watany-dev @yusukebe
I removed feature for switching SSR/SSG.
I'll make other PR to add API like `isSSG`.

File: deno_dist/helper/ssg/index.ts [hunk 2]
```diff
@@ -40,19 +64,29 @@ export const generateHtmlMap = async <
   S extends Schema = {},
   BasePath extends string = '/'
 >(
-  app: Hono<E, S, BasePath>
+  app: Hono<E, S, BasePath>,
+  options: ToSSGOptions
```
Comment (yusukebe) [inline deno_dist/helper/ssg/index.ts:68]: You can remove this `options`.

**Review by yusukebe — COMMENTED** [review]
Hi @nakasyou 

Please check the comment.

**Comment by watany-dev** [comment]
Speaking of which, have you decided on ssgParams as the name of the middleware? I feel like there is a name that is more in line with the actual situation than the original multiple function params, and it may be okay to keep it as it is.

**Comment by yusukebe** [comment]
ssgPrams is middleware, but it is exported from `hono/ssg`, so I think "**SSG Helper**" should be the "common name".

File: deno_dist/helper/ssg/index.ts (hunk 2 above)
Comment (nakasyou) [reply deno_dist/helper/ssg/index.ts]: I removed it.

File: src/helper/ssg/index.test.tsx [hunk 3]
```diff
@@ -30,20 +30,28 @@ describe('toSSG function', () => {
...
+      }
+    ]), c => c.html(<h1>{c.req.param('post')}</h1>))
 
     const fsMock = {
       writeFile: vi.fn((path, data) => Promise.resolve()),
       mkdir: vi.fn((path, options) => Promise.resolve()),
     }
 
-    const htmlMap = await generateHtmlMap(app)
+    const htmlMap = await generateHtmlMap(app, {})
```
Comment (yusukebe) [inline src/helper/ssg/index.test.tsx:48]: We should write the tests for checking the contents of `htmlMap`. Like this:

```ts
expect(htmlMap.get('/post/1')).toBe('<h1>1</h1>')
```

**Comment by yusukebe** [comment]
Hi @nakasyou 

I've left the comment about testing. Please check it and resolve the conflicts.

File: src/helper/ssg/index.test.tsx (hunk 3 above)
Comment (nakasyou) [reply src/helper/ssg/index.test.tsx]: I did it.

**Comment by nakasyou** [comment]
@yusukebe
Thank you for review.
I added that test code, and resolve conflicts.

**Comment by yusukebe** [comment]
@nakasyou 

Thanks! Let's land it.

**Pull request merged**